pkg/controller/statefulset/stateful_set_control.go

```diff
@@ -132,34 +308,58 @@ func (ssc *defaultStatefulSetControl) UpdateStatefulSet(set *apps.StatefulSet, p
 		if isFailed(replicas[i]) {
 			glog.V(2).Infof("StatefulSet %s is recreating failed Pod %s", set.Name, replicas[i].Name)
 			if err := ssc.podControl.DeleteStatefulPod(set, replicas[i]); err != nil {
-				return err
+				return &status, err
+			}
+			if getPodRevision(pods[i]) == currentRevision.Name {
+				status.CurrentReplicas--
+			} else if getPodRevision(pods[i]) == updateRevision.Name {
+				status.UpdatedReplicas--
 			}
-			replicas[i] = newStatefulSetPod(set, i)
+			status.Replicas--
+			replicas[i] = newVersionedStatefulSetPod(
+				currentSet,
+				updateSet,
+				currentRevision.Name,
+				updateRevision.Name,
+				i)
 		}
 		// If we find a Pod that has not been created we create the Pod
 		if !isCreated(replicas[i]) {
 			if err := ssc.podControl.CreateStatefulPod(set, replicas[i]); err != nil {
-				return err
+				return &status, err
 			}
+			status.Replicas++
+			if getPodRevision(replicas[i]) == currentRevision.Name {
+				status.CurrentReplicas++
+			} else if getPodRevision(replicas[i]) == updateRevision.Name {
+				status.UpdatedReplicas++
+			}
+
 			// if the set does not allow bursting, return immediately
 			if monotonic {
-				return nil
+				return &status, nil
 			}
 			// pod created, no more work possible for this round
 			continue
 		}
 		// If we find a Pod that is currently terminating, we must wait until graceful deletion
 		// completes before we continue to make progress.
 		if isTerminating(replicas[i]) && monotonic {
-			glog.V(2).Infof("StatefulSet %s is waiting for Pod %s to Terminate", set.Name, replicas[i].Name)
-			return nil
+			glog.V(2).Infof(
+				"StatefulSet %s is waiting for Pod %s to Terminate",
```

Comment:
All debug in this controller should be at v(4) or above.  Only truly exceptional cases should be below v(4).  Errors should be utilruntime.HandleError()